active/002-py3-only.md

```diff
@@ -0,0 +1,72 @@
+---
+Feature name: py3-only
+Start date: 2015-09-23
+Pull request: 6
+Authors: ["@jairideout", "@gregcaporaso"]
+Contributors:
+-
+---
+
+# Summary
+
+Remove Python 2 (Legacy Python) support from scikit-bio in favor of Python 3
+support only.
+
+# Motivation
+
+We propose removing Python 2 (Legacy Python) support from scikit-bio for the
+following reasons:
+
+1. The scientific Python community is in favor of dropping support for Python 2
+as soon as possible (e.g., see [recent discussion at DS4DS](https://bids.hackpad.com/Planning-an-early-death-for-Python-2-Legacy-Python.-6aB7agi5XCr)).
+2. Dropping Python 2 support will reduce development and maintenance burden on
+scikit-bio developers.
+3. Supporting Python 3 only will allow us to start using new features of the
+language, ultimately making scikit-bio a better package.
+4. Given that scikit-bio is still in beta and has relatively few users (e.g.,
+compared to a package like numpy), making this switch now would be better than
+waiting and could set a precedent for other scientific Python packages.
+
+# Detailed design
+
+Changes to support Python 3-only include:
+
+1. Remove Python 2 tests from Travis-CI build matrix.
+2. Remove `future` and `six` dependencies from scikit-bio.
+3. Update documentation to indicate Python 3-only, including (but not limited
+to) `README.md` and `CONTRIBUTING.md`.
+4. Remove `doc/source/development/py3.rst`.
+5. Update PyPI classifiers in `setup.py` to indicate Python 3-only.
+6. Search codebase for references to Python 2 and remove/update accordingly.
+This includes comments and actual code modifications.
+7. Removes hacks in `skbio.io` necessary for Python 2 compatibility.
+8. Test against latest release and nightly build of Python 3 on Travis-CI.
+Allow failure against nightly Python 3 build.
+
+Note: scikit-bio's doctests are already Python 3-only, so no updates necessary
+there! :shipit:
+
+# Drawbacks
+
+Packages/tools supporting Python 2 (including those supporting Python 2
+**and** 3) will not be able to depend on scikit-bio.
+
+# Alternatives
+
+There are no feasible alternatives (that we know of) given our limited
+developer bandwidth and free (as in :beer:) software model. Therefore, this
+transition **must** happen at some point because Python 2 will eventually not
+be supported by other systems and tools. By not making the switch now (or
+soonish), we are delaying the inevitable and will affect more users as our
+userbase increases in size. It would be good form to make this change now
+while we are still in beta.
+
+# Unresolved questions
+
+What is a reasonable timeline for this change? Can we do this now or should we
+plan this for a release that is, for example, one year away?
+
+Are there scikit-bio users who would stop using the package if it were Python
+3 only?
+
+Would Python 3-only support prevent the adoption of scikit-bio by new users?
```

Comment:
Perhaps this should be a *big* banner in the front page that says
something along the lines of "scikit-bio only supports Python 3 - read
this document to find out why", hopefully avoiding "surprises" for
newcomers.

That and we can perhaps start a big media campaign - maybe Obama would
like to participate now that he is leaving the office. "Whenever I
process large sequence files, I do this with scikit-bio using PY3K - I'm
Barack Obama and I approve this message".

On (Oct-05-15|11:46), Jai Ram Rideout wrote:
> > +developer bandwidth and free (as in :beer:) software model. Therefore, this
> > +transition **must** happen at some point because Python 2 will eventually not
> > +be supported by other systems and tools. By not making the switch now (or
> > +soonish), we are delaying the inevitable and will affect more users as our
> > +userbase increases in size. It would be good form to make this change now
> > +while we are still in beta.
> > +
> > +# Unresolved questions
> > +
> > +What is a reasonable timeline for this change? Can we do this now or should we
> > +plan this for a release that is, for example, one year away?
> > +
> > +Are there scikit-bio users who would stop using the package if it were Python
> > +3 only?
> > +
> > +Would Python 3-only support prevent the adoption of scikit-bio by new users?
> 
> I agree. Hopefully by updating our user-facing docs (point 3 in "Detailed design" ) we can better inform and educate our users of this.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio-rfcs/pull/6/files#r41181345